src/spider/client/Context.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef SPIDER_CLIENT_CONTEXT_HPP
+#define SPIDER_CLIENT_CONTEXT_HPP
+
+#include <boost/uuid/uuid.hpp>
+#include <functional>
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "../core/Serializer.hpp"
+#include "Data.hpp"
+#include "Job.hpp"
+#include "TaskGraph.hpp"
+
+namespace spider {
+class ContextImpl;
+
+class Context {
+public:
+    /**
+     * Aborts the current running task and job. This function never returns.
+     *
+     * @param message Error message indicating the reason for the abort.
+     */
+    auto abort(std::string const& message);
+
+    /**
+     * @return ID of the current running task instance.
+     */
+    [[nodiscard]] auto get_id() const -> boost::uuids::uuid;
+
+    /**
+     * Gets data by key.
+     *
+     * NOTE: Callers cannot get data created by other tasks, but they can get data created by
+     * previous instances of the same task.
+     *
+     * @tparam Value
+     * @param key
+     * @return An optional containing the data if the given key exists, or `std::nullopt` otherwise.
+     */
+    template <Serializable Value>
+    auto get_data(std::string const& key) -> std::optional<Data<Value>>;
+
+    /**
+     * Inserts the given key-value pair into the key-value store, overwriting any existing value.
+     *
+     * @param key
+     * @param value
+     */
+    auto insert_kv(std::string const& key, std::string const& value);
+
+    /**
+     * Gets the value corresponding to the given key.
+     *
+     * NOTE: Callers cannot get values created by other tasks, but they can get values created by
+     * previous instances of the same task.
+     *
+     * @param key
+     * @return An optional containing the value if the given key exists, or `std::nullopt`
+     * otherwise.
+     */
+    auto get_kv(std::string const& key) -> std::optional<std::string>;
```

Comment:
I think `get_key_value` might be a better name

Reply:
`get_key_value` sounds like it's going to return both the key and value. It would be similar to calling `get_data` as `get_key_data`. I think `get_value` is clear since that's the return value and the parameter indicates that you need to pass in the `key.

Reply:
I think this is special because key-value store is a widely used jargon. If the same logic applies, how should we call `insert_kv`. Since it accepts a key and a value, then should it just be called `insert`?

Reply:
I get what you're saying but I still don't think `get_key_value` is clear. Fundamentally, the issue is that `get_kv` and `get_data` are operations on two different data stores instead of operations on the Driver/Context. How about calling the two methods `kv_store_get` and `data_store_get`, respectively?